File: golang/linked_list/max_twin_sum_of_linked_list.go
package main

func reverseList(head *ListNode) *ListNode {
	var prev *ListNode

	curr := head

	for curr != nil {
		next := curr.Next

		curr.Next = prev

		prev = curr

		curr = next
	}

	return prev
}

func pairSum(head *ListNode) int {
	slow, fast := head, head

	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	secondHalf := reverseList(slow)

	maxSum := 0

	firstHalf := head

	for secondHalf != nil {
		twinSum := firstHalf.Val + secondHalf.Val

		if twinSum > maxSum {
			maxSum = twinSum
		}

		firstHalf = firstHalf.Next
		secondHalf = secondHalf.Next
	}

	return maxSum
}

File: java/src/main/java/com/sanedge/interview/linked_list/MaxTwinSumOfLinkedList.java
package com.sanedge.interview.linked_list;

import com.sanedge.interview.hashmap.TwoSum;

public class MaxTwinSumOfLinkedList {
  public ListNode reverseList(ListNode head) {
    ListNode prev = null;

    ListNode curr = head;

    while (curr != null) {
      ListNode next = curr.next;

      curr.next = prev;

      prev = curr;

      curr = next;
    }

    return prev;
  }

  public int pairSum(ListNode head) {
    ListNode slow = head;
    ListNode fast = head;

    while (fast != null && fast.next != null) {
      slow = slow.next;
      fast = fast.next.next;
    }

    ListNode secondHalf = reverseList(slow);

    int maxSum = 0;
    ListNode firstHalf = head;

    while (secondHalf != null) {
      int twinSum = firstHalf.val + secondHalf.val;

      if (twinSum > maxSum) {
        maxSum = twinSum;
      }

      firstHalf = firstHalf.next;
      secondHalf = secondHalf.next;
    }

    return maxSum;

  }

}
